File: music-android/app/src/main/java/com/sismics/music/util/CacheUtil.java
package com.sismics.music.util;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.util.Log;

import com.sismics.music.model.Album;
import com.sismics.music.model.Artist;
import com.sismics.music.model.FullAlbum;
import com.sismics.music.model.PlaylistTrack;
import com.sismics.music.model.Track;
import com.snappydb.DB;
import com.snappydb.SnappyDB;
import com.snappydb.SnappydbException;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Cache utilities.
 *
 * @author bgamard.
 */
public class CacheUtil {
    /**
     * Returns the music cache directory.
     * @return Music cache directory
     */
    public static File getMusicCacheDir(Context context) {
        File cacheDir = ContextCompat.getExternalCacheDirs(context)[0];
        File cache = new File(cacheDir, "music");
        if (!cache.exists()) {
            cache.mkdirs();
        }
        return cache;
    }

    /**
     * Returns true if the given track is complete.
     * @param trackId Track ID
     * @return True if complete
     */
    public static boolean isTrackCached(Context context, String trackId) {
        try {
            return SnappyDB.with(context).exists("track:" + trackId);
        } catch (SnappydbException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Returns true if the given track is complete.
     * @param albumId Album ID
     * @return True if complete
     */
    public static boolean isAlbumCached(Context context, String albumId) {
        try {
            return SnappyDB.with(context).exists("album:" + albumId);
        } catch (SnappydbException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Returns the complete playlistTrack file.
     * @param playlistTrack PlaylistTrack
     * @return Complete playlistTrack file
     */
    public static File getCompleteCacheFile(Context context, PlaylistTrack playlistTrack) {
        return new File(getMusicCacheDir(context), playlistTrack.getArtist().getId()
                + File.separator + playlistTrack.getAlbum().getId()
                + File.separator + playlistTrack.getTrack().getId() + ".complete");
    }

    public static File getCompleteCacheFile(Context context, String artistId, String albumId, String trackId) {
        return new File(getMusicCacheDir(context), artistId
                + File.separator + albumId
                + File.separator + trackId + ".complete");
    }

    /**
     * Returns the incomplete playlistTrack file.
     * @param playlistTrack PlaylistTrack
     * @return Incomplete playlistTrack file
     */
    public static File getIncompleteCacheFile(Context context, PlaylistTrack playlistTrack) {
        File albumDir = new File(getMusicCacheDir(context), playlistTrack.getArtist().getId() + File.separator + playlistTrack.getAlbum().getId());
        if (!albumDir.exists()) {
            albumDir.mkdirs();
        }
        return new File(albumDir, playlistTrack.getTrack().getId());
    }

    /**
     * Set a track as complete.
     * @param context Context
     * @param playlistTrack PlaylistTrack
     * @param file File
     * @return True if the operation was successful
     */
    public static boolean setComplete(Context context, PlaylistTrack playlistTrack, File file) {
        try {
            DB snappyDb = SnappyDB.with(context);
            snappyDb.put("album:" + playlistTrack.getAlbum().getId(), playlistTrack.getAlbum());
            snappyDb.put("artist:" + playlistTrack.getArtist().getId(), playlistTrack.getArtist());
            snappyDb.put("track:" + playlistTrack.getTrack().getId(), playlistTrack.getTrack());
        } catch (SnappydbException e) {
            throw new RuntimeException(e);
        }
        return file.renameTo(new File(file.getAbsolutePath() + ".complete"));
    }

    /**
     * Returns cached tracks for an album.
     * @param album Album
     * @return Cached tracks
     */
    public static List<Track> getCachedTrackList(Context context, Artist artist, Album album) {
        List<Track> trackList = new ArrayList<>();
        File albumDir = new File(getMusicCacheDir(context), artist.getId() + File.separator + album.getId());
        if (!albumDir.exists()) {
            return trackList;
        }

        // List complete files from this album
        File[] files = albumDir.listFiles((dir, filename) -> filename.endsWith(".complete"));

        // Get tracks from cache
        try {
            DB snappyDb = SnappyDB.with(context);
            for (File file : files) {
                String trackKey = "track:" + file.getName().substring(0, file.getName().indexOf("."));
                if (snappyDb.exists(trackKey)) {
                    trackList.add(snappyDb.get(trackKey, Track.class));
                }
            }
        } catch (SnappydbException e) {
            Log.e("CacheUtil", "Error getting tracks from cache", e);
        }

        return trackList;
    }

    /**
     * Returns cached albums.
     * @return Albums
     */
    public static List<FullAlbum> getCachedAlbumList(Context context) {
        List<FullAlbum> albumList = new ArrayList<>();
        try {
            DB snappyDb = SnappyDB.with(context);
            for (String albumKey : SnappyDB.with(context).findKeys("album:")) {
                Album album = snappyDb.get(albumKey, Album.class);
                Artist artist = snappyDb.get("artist:" + album.getArtistId(), Artist.class);
                albumList.add(new FullAlbum(artist, album));
            }
        } catch (SnappydbException e) {
            throw new RuntimeException(e);
        }
        return albumList;
    }

    /**
     * Remove a track from the cache.
     * @param artistId Artist ID
     * @param albumId Album ID
     * @param trackId Track ID
     */
    public static void removeTrack(Context context, String artistId, String albumId, String trackId) {
        File file = getCompleteCacheFile(context, artistId, albumId, trackId);
        if (file.exists()) {
            file.delete();
        }

        try {
            DB snappyDb = SnappyDB.with(context);
            String trackKey = "track:" + trackId;
            String albumKey = "album:" + albumId;
            String artistKey = "artist:" + artistId;
            if (snappyDb.exists(trackKey)) {
                snappyDb.del(trackKey);
            }

            File albumDir = new File(getMusicCacheDir(context), artistId + File.separator + albumId);
            if (albumDir.exists()) {
                File[] trackFileList = albumDir.listFiles();
                boolean empty = true;
                for (File trackFile : trackFileList) {
                    if (trackFile.getName().endsWith(".complete")) {
                        empty = false;
                    } else {
                        trackFile.delete();
                    }
                }

                if (empty) {
                    albumDir.delete();
                    if (snappyDb.exists(albumKey)) {
                        snappyDb.del(albumKey);
                    }
                }
            }

            File artistDir = new File(getMusicCacheDir(context), artistId);
            if (artistDir.exists() && artistDir.list().length == 0) {
                artistDir.delete();
                if (snappyDb.exists(artistKey)) {
                    snappyDb.del(artistKey);
                }
            }
        } catch (SnappydbException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Remove an album from the cache.
     * @param artistId Artist ID
     * @param albumId Album ID
     */
    public static void removeAlbum(Context context, String artistId, String albumId) {
        File albumDir = new File(getMusicCacheDir(context), artistId + File.separator + albumId);
        if (albumDir.exists()) {
            File[] fileList = albumDir.listFiles((dir, filename) -> filename.endsWith(".complete"));
            for (File file : fileList) {
                String trackId = file.getName().substring(0, file.getName().indexOf("."));
                removeTrack(context, artistId, albumId, trackId);
            }
        }
    }
}

File: music-android/app/src/main/java/com/sismics/music/util/CacheUtil.kt
package com.sismics.music.util

import android.os.Environment
import android.util.Log
import com.sismics.music.model.Album
import com.sismics.music.model.PlaylistTrack
import com.sismics.music.model.Track
import org.jaudiotagger.audio.mp3.MP3FileReader
import org.jaudiotagger.tag.FieldKey
import java.io.File
import java.nio.channels.ClosedChannelException
import java.util.*

/**
 * Cache utilities.
 *
 * @author bgamard.
 */
object CacheUtil {
    private val TAG = CacheUtil.javaClass.simpleName

    /**
     * Returns the music cache directory.
     *
     * @return Music cache directory
     */
    val musicCacheDir: File
        get() {
            val music = Environment.getExternalStoragePublicDirectory("SismicsMusic")
            val cache = File(music, "cache")
            if (!cache.exists()) {
                cache.mkdirs()
            }
            return cache
        }

    /**
     * Returns true if a playlistTrack is complete.
     *
     * @param playlistTrack PlaylistTrack
     * @return True if complete
     */
    fun isComplete(playlistTrack: PlaylistTrack): Boolean {
        return getCompleteCacheFile(playlistTrack).exists()
    }

    /**
     * Returns true if the given track with album is complete.
     *
     * @param album Album associated with the track
     * @param track Track
     * @return True if complete
     */
    fun isComplete(album: Album, track: Track): Boolean {
        val playlistTrack = PlaylistTrack(album, track)
        return getCompleteCacheFile(playlistTrack).exists()
    }

    /**
     * Returns the complete playlistTrack file.
     *
     * @param playlistTrack PlaylistTrack
     * @return Complete playlistTrack file
     */
    fun getCompleteCacheFile(playlistTrack: PlaylistTrack): File {
        val albumDir = File(musicCacheDir, playlistTrack.albumId)
        if (!albumDir.exists()) {
            albumDir.mkdirs()
        }
        return File(albumDir, playlistTrack.id + ".complete")
    }

    /**
     * Returns the incomplete playlistTrack file.
     *
     * @param playlistTrack PlaylistTrack
     * @return Incomplete playlistTrack file
     */
    fun getIncompleteCacheFile(playlistTrack: PlaylistTrack): File {
        val albumDir = File(musicCacheDir, playlistTrack.albumId)
        if (!albumDir.exists()) {
            albumDir.mkdirs()
        }
        return File(albumDir, playlistTrack.id)
    }

    /**
     * Set a track as complete.
     *
     * @param file File
     * @return True if the operation was successful
     */
    fun setComplete(file: File): Boolean {
        return file.renameTo(File(file.absolutePath + ".complete"))
    }

    /**
     * Returns cached tracks for an album.
     *
     * @param album Album
     * @return Cached tracks
     */
    fun getCachedTrack(album: Album): List<Track> {
        val trackList = ArrayList<Track>()
        val albumDir = File(musicCacheDir, album.id)
        if (!albumDir.exists()) {
            return trackList
        }

        // List complete files from this album
        val files = albumDir.listFiles { dir, filename -> filename.endsWith(".complete") }

        // Extract tags from cached files
        for (file in files) {
            try {
                val audioFile = MP3FileReader().read(file)
                val tag = audioFile.tag
                val header = audioFile.audioHeader

                val id = file.name.substring(0, file.name.indexOf("."))
                val title = tag?.getFirst(FieldKey.TITLE)
                val track = Track(id, title?:"", header.trackLength)

                trackList.add(track)
            } catch (e: ClosedChannelException) {
                // We have been interrupted, don't care
            } catch (e: Exception) {
                Log.e(TAG, "Error extracting metadata from file: " + file.absolutePath, e)
            }
        }

        return trackList
    }

    /**
     * Returns album IDs containing at least one cached track.
     *
     * @return Albums IDs
     */
    val cachedAlbumSet: Set<String>
        get() {
            val cacheDir = musicCacheDir
            val albumList = cacheDir.listFiles()
            val output = HashSet<String>()
            for (album in albumList) {
                if (album.list { dir, filename -> filename.endsWith(".complete") }.isNotEmpty()) {
                    output.add(album.name)
                }
            }
            return output
        }

    /**
     * Remove a track from the cache.
     *
     * @param album Album
     * @param track Track
     */
    fun removeTrack(album: Album, track: Track) {
        val playlistTrack = PlaylistTrack(album, track)
        val file = getCompleteCacheFile(playlistTrack)
        if (file.exists()) {
            file.delete()
        }
    }

    /**
     * Remove an album from the cache.
     *
     * @param albumId Album ID
     */
    fun removeAlbum(albumId: String) {
        val albumDir = File(musicCacheDir, albumId)
        if (albumDir.exists()) {
            for (file in albumDir.listFiles()) {
                file.delete()
            }
            albumDir.delete()
        }
    }
}
